site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+	c.App.SetContext(ctx)
+	// ...
+	// Execute the actual API handler
+	h.HandleFunc(c, w, r)
+}
+```
+
+Next, we'll modify the actual API handler to nest it inside the parent span (we'll use `SearchUsers` as example)
+
+```go
+func (a *App) SearchUsers(props *model.UserSearch, options *model.UserSearchOptions) ([]*model.User, *model.AppError) {
+	// save previous context
+	origCtx := a.ctx
+	// generate new span, nested inside the parent span
+	span, newCtx := tracing.StartSpanWithParentByContext(a.ctx, "app.SearchUsers")
+	// set new context
+	a.ctx = newCtx
+	
+	// log some parameters
+	span.SetTag("searchProps", props)
+
+	// on function exit, restore context and finish the span
+	defer func() {
+		a.ctx = origCtx
+		span.Finish()
+	}()
+
+	// ...
+	// perform actual work
+	// ...
+
+	// in case of an error, add it to the span
+	if err != nil {
+		span.LogFields(spanlog.Error(err))
+		ext.Error.Set(span, true)
+	}
+
+	// return results
+}
+```
+
+Rather straight forward, right? We marked our "entry-point" by creating a root span, populating it with useful context information, then passing the context down the stack and creating new span underneath it.
+
+We could stop right here, because this is all you need to have a working trace! **But**: for a large application like `mattermost-server` wrapping all of the 900+ API handlers in tracing code would be incredible labor intensive AND will create a lot of noise in the source code.
+
+So, can we do better?
+
+## Decorator pattern
+
+Before diving into our solution, I want to first introduce Decorator Pattern.
+
+To quote [Wikipedia](https://en.wikipedia.org/wiki/Decorator_pattern):
+
+> In object-oriented programming, the decorator pattern is a design pattern that allows behavior to be added to an individual object, dynamically, without affecting the behavior of other objects from the same class. The decorator pattern is often useful for adhering to the Single Responsibility Principle, as it allows functionality to be divided between classes with unique areas of concern. The decorator pattern is structurally nearly identical to the chain of responsibility pattern, the difference being that in a chain of responsibility, exactly one of the classes handles the request, while for the decorator, all classes handle the request.
+
+![decorator pattern](https://upload.wikimedia.org/wikipedia/commons/8/83/W3sDesign_Decorator_Design_Pattern_UML.jpg?1591628685344)
+
+In simpler terms, let's say we have an object called `Cow` that has some methods:
+
+![cow undecorated](/blog/2020-06-10-opentracing/cowundecorated.png) 
+
+We want to introduce additional functionality on top of what `Cow` already does, without modifying the actual code of the `Cow` object. For example, we want to measure performance of each method, and log the parameters that are being passed to each method. Here's how it would look if we apply the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern):
```

Comment:
```suggestion
We want to introduce additional functionality on top of what `Cow` already does, without modifying the actual code of the `Cow` object. For example, we want to measure performance of each method and log the parameters that are being passed to each method. Here's how it would look if we apply the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern):
```